diff --git a/asmcomp/arm64/emit.mlp b/asmcomp/arm64/emit.mlp
index 1dbb4b2b1ead..a119a5b67f87 100644
--- a/asmcomp/arm64/emit.mlp
+++ b/asmcomp/arm64/emit.mlp
@@ -513,10 +513,7 @@ module BR = Branch_relaxation.Make (struct
     | Lop (Ispecific Imove32) -> 1
     | Lop (Ispecific (Isignext _)) -> 1
     | Lop (Idls_get) -> 1
-    | Lop (Ireturn_addr) ->
-        invalid_arg (
-          "Support for Ireturn_addr is not implemented on architecture "
-          ^ Config.architecture)
+    | Lop (Ireturn_addr) -> 1
     | Lreloadretaddr -> 0
     | Lreturn -> epilogue_size f
     | Llabel _ -> 0
@@ -962,9 +959,11 @@ let emit_instr env i =
         let offset = Domainstate.(idx_of_field Domain_dls_root) * 8 in
         `	ldr	{emit_reg i.res.(0)}, [{emit_reg reg_domain_state_ptr}, {emit_int offset}]\n`
     | Lop(Ireturn_addr) ->
-        invalid_arg (
-          "Support for Ireturn_addr is not implemented on architecture "
-          ^ Config.architecture)
+        let n = frame_size env in
+        if env.f.fun_frame_required then
+          `	ldr	{emit_reg i.res.(0)}, [sp, #{emit_int (n-8)}]\n`
+        else
+          `	mov	{emit_reg i.res.(0)}, x30\n`
     | Lreloadretaddr ->
         ()
     | Lreturn ->
