diff --git a/asmcomp/power/emit.mlp b/asmcomp/power/emit.mlp
index 9d7582933849..47f5419a92c0 100644
--- a/asmcomp/power/emit.mlp
+++ b/asmcomp/power/emit.mlp
@@ -447,10 +447,7 @@ module BR = Branch_relaxation.Make (struct
     | Lop(Iopaque) -> 0
     | Lop(Ispecific _) -> 1
     | Lop (Idls_get) -> 1
-    | Lop (Ireturn_addr) ->
-        invalid_arg (
-          "Support for Ireturn_addr is not implemented on architecture "
-          ^ Config.architecture)
+    | Lop (Ireturn_addr) -> 1
     | Lreloadretaddr -> 2
     | Lreturn -> 2
     | Llabel _ -> 0
@@ -839,9 +836,10 @@ let emit_instr env i =
         let offset = Domainstate.(idx_of_field Domain_dls_root) * 8 in
         `	ld	{emit_reg i.res.(0)}, {emit_int offset}(30)\n`
     | Lop (Ireturn_addr) ->
-        invalid_arg (
-          "Support for Ireturn_addr is not implemented on architecture "
-          ^ Config.architecture)
+        if env.f.fun_frame_required then
+        `	ld	{emit_reg i.res.(0)}, {emit_int(retaddr_offset env)}(1)\n`
+        else
+        `	mflr	{emit_reg i.res.(0)}\n`
     | Lreloadretaddr ->
         `	ld	11, {emit_int(retaddr_offset env)}(1)\n`;
         `	mtlr	11\n`
