diff --git a/asmcomp/s390x/emit.mlp b/asmcomp/s390x/emit.mlp
index c98a0d77623f..2d6bc4d1dc30 100644
--- a/asmcomp/s390x/emit.mlp
+++ b/asmcomp/s390x/emit.mlp
@@ -595,9 +595,11 @@ let emit_instr env i =
         let ofs = Domainstate.(idx_of_field Domain_dls_root) * 8 in
         `	lg	{emit_reg i.res.(0)}, {emit_int ofs}(%r10)\n`
     | Lop (Ireturn_addr) ->
-        invalid_arg (
-          "Support for Ireturn_addr is not implemented on architecture "
-          ^ Config.architecture)
+        let n = frame_size env in
+        if env.f.fun_frame_required then
+            `	lg	{emit_reg i.res.(0)}, {emit_int(n - size_addr)}(%r15)\n`
+        else
+            `	lgr	{emit_reg i.res.(0)}, %r14\n`
     | Lreloadretaddr ->
         let n = frame_size env in
         `	lg	%r14, {emit_int(n - size_addr)}(%r15)\n`
