diff --git a/asmcomp/emitaux.ml b/asmcomp/emitaux.ml
index 306e24922399..61ec3d940f44 100644
--- a/asmcomp/emitaux.ml
+++ b/asmcomp/emitaux.ml
@@ -45,7 +45,7 @@ let emit_symbol s =
       if c = Compilenv.symbol_separator then
         output_char !output_channel c
       else
-        Printf.fprintf !output_channel "$%02x" (Char.code c)
+        Printf.fprintf !output_channel "$$%02x" (Char.code c)
   done
 
 let emit_string_literal s =
diff --git a/asmcomp/x86_proc.ml b/asmcomp/x86_proc.ml
index b7abe4d3d1ca..349a22e5a624 100644
--- a/asmcomp/x86_proc.ml
+++ b/asmcomp/x86_proc.ml
@@ -82,12 +82,8 @@ let string_of_symbol prefix s =
     Buffer.add_string b prefix;
     String.iter
       (fun c ->
-       (* FIXME: using $ to prefix escaped characters can make names
-          ambiguous if the symbol separator is also set to $; a different
-          escape prefix should be used in this case, if this ever causes
-          problems in the real world. *)
        if is_special_char c then
-         Printf.bprintf b "$%02x" (Char.code c)
+         Printf.bprintf b "$$%02x" (Char.code c)
        else
          Buffer.add_char b c
       )
diff --git a/middle_end/compilenv.ml b/middle_end/compilenv.ml
index d975c1781ce6..8ee77bc34382 100644
--- a/middle_end/compilenv.ml
+++ b/middle_end/compilenv.ml
@@ -90,10 +90,7 @@ let current_unit =
     ui_export_info = default_ui_export_info;
     ui_for_pack = None }
 
-let symbol_separator =
-  match Config.ccomp_type with
-  | "msvc" -> '$' (* MASM does not allow for dots in symbol names *)
-  | _ -> '.'
+let symbol_separator = '$'
 
 let concat_symbol unitname id =
   Printf.sprintf "%s%c%s" unitname symbol_separator id
diff --git a/testsuite/tests/asmcomp/func_sections.run b/testsuite/tests/asmcomp/func_sections.run
index cb63628bcd2c..8a0038b0b789 100755
--- a/testsuite/tests/asmcomp/func_sections.run
+++ b/testsuite/tests/asmcomp/func_sections.run
@@ -7,4 +7,4 @@ ${program}
 
 # now check the assembly file produced during compilation
 asm=${test_build_directory}/func_sections.s
-grep "\.section \.text\.caml\.camlFunc_sections\." "$asm" | wc -l | tr -d ' ' | sed '/^$/d'
+grep "\.section \.text\.caml\.camlFunc_sections\\$" "$asm" | wc -l | tr -d ' ' | sed '/^$/d'
