diff --git a/asmcomp/arm64/emit.mlp b/asmcomp/arm64/emit.mlp
index db259e68b447..baee6746caa9 100644
--- a/asmcomp/arm64/emit.mlp
+++ b/asmcomp/arm64/emit.mlp
@@ -765,7 +765,7 @@ let emit_instr env i =
           (* Store OCaml stack in x19 register and restore later. *)
           `	mov	x19, sp\n`;
           cfi_remember_state ();
-          cfi_def_cfa_register ~reg:29;
+          cfi_def_cfa_register ~reg:19;
           let offset = Domainstate.(idx_of_field Domain_c_stack) * 8 in
           `	ldr	{emit_reg reg_tmp1}, [{emit_reg reg_domain_state_ptr}, {emit_int offset}]\n`;
           `	mov	sp, {emit_reg reg_tmp1}\n`;
