diff --git a/asmcomp/asmlibrarian.ml b/asmcomp/asmlibrarian.ml
index 5b1c315bb283..823b07db3a8d 100644
--- a/asmcomp/asmlibrarian.ml
+++ b/asmcomp/asmlibrarian.ml
@@ -68,7 +68,7 @@ let create_archive file_list lib_name =
          (fun (filename, (unit, _crc)) ->
             Linkdeps.add ldeps
               ~filename ~compunit:unit.ui_name
-              ~provides:unit.ui_defines
+              ~provides:[unit.ui_name]
               ~requires:(List.map fst unit.ui_imports_cmx))
          (List.rev units);
        (match Linkdeps.check ldeps with
diff --git a/asmcomp/asmlink.ml b/asmcomp/asmlink.ml
index ce021c1496ea..e8179a729f7c 100644
--- a/asmcomp/asmlink.ml
+++ b/asmcomp/asmlink.ml
@@ -156,7 +156,7 @@ let scan_file ldeps file tolink = match file with
       (* This is a .cmx file. It must be linked in any case. *)
       Linkdeps.add ldeps
         ~filename:file_name ~compunit:info.ui_name
-        ~provides:info.ui_defines
+        ~provides:[info.ui_name]
         ~requires:(List.map fst info.ui_imports_cmx);
       (info, file_name, crc) :: tolink
   | Library (file_name,infos) ->
@@ -171,7 +171,7 @@ let scan_file ldeps file tolink = match file with
            then begin
              Linkdeps.add ldeps
                ~filename:file_name ~compunit:info.ui_name
-               ~provides:info.ui_defines
+               ~provides:[info.ui_name]
                ~requires:(List.map fst info.ui_imports_cmx);
              (info, file_name, crc) :: reqd
            end else
