diff --git a/Changes b/Changes
index 160793009187..458284cc673c 100644
--- a/Changes
+++ b/Changes
@@ -45,6 +45,10 @@ _______________
   bytecode linker in `-custom` and `-output-*` modes.
   (David Allsopp, Antonin Décimo and Samuel Hym, review by Vincent Laviron)
 
+- #xxxxx: Slightly change the s390x assembly dialect in order to build with
+  Clang's integrated assembler.
+  (Miod Vallat, review by TBD)
+
 ### Build system:
 
 ### Bug fixes:
diff --git a/runtime/s390x.S b/runtime/s390x.S
index 5b7cdc8e4a88..52a5e228b52e 100644
--- a/runtime/s390x.S
+++ b/runtime/s390x.S
@@ -35,7 +35,7 @@
 
 #define LBL(x) .L##x
 #define G(r) r
-#define GREL(r) r@GOTENT
+#define GREL(r) r@GOT
 #define GCALL(r) r@PLT
 #define TEXT_SECTION(name)
 #define FUNCTION(name) \
