diff --git a/Changes b/Changes
index 4c20cab050fa..b4cdc5d6f8b6 100644
--- a/Changes
+++ b/Changes
@@ -195,6 +195,10 @@ _______________
 - #13162: Use quoted strings to clarify code being generated.
   (Antonin Décimo, review by Miod Vallat and Gabriel Scherer)
 
+- #13015: Emit floating-point literals in .rodata section on ELF arm64
+  platforms (Linux, *BSD).
+  (Miod Vallat, review by Nicolás Ojeda Bär)
+
 ### Build system:
 
 - #12909: Reorganise how MKEXE_VIA_CC is built to make it correct for MSVC by
