diff --git a/Changes b/Changes
index 1662f7366992..ede543acf907 100644
--- a/Changes
+++ b/Changes
@@ -190,6 +190,10 @@ Working version
 - #13541, #13591: Fix headers for C++ inclusion.
   (Antonin Décimo, review by Nick Barnes, report by Kate Deplaix)
 
+- #13595: Use x19 as Canonical Frame Address (CFA) register. This would cause
+  backtraces to be truncated when calling no alloc C code.
+  (Tim McGilchrist, report by Nick Barnes, review by Nick Barnes)
+
 OCaml 5.3.0
 ___________
 
diff --git a/testsuite/tests/unwind/check-linker-version.sh b/testsuite/tests/unwind/check-linker-version.sh
index fabd0ed9bc11..1966885da19c 100755
--- a/testsuite/tests/unwind/check-linker-version.sh
+++ b/testsuite/tests/unwind/check-linker-version.sh
@@ -1,13 +1,14 @@
-#!/bin/sh
-exec > ${ocamltest_response} 2>&1
-LDFULL="`ld -v 2>&1`"
-LD="`echo $LDFULL | grep -o \"ld64-[0-9]*\"`"
-LDVER="`echo $LD | sed \"s/ld64-//\"`"
-if [[ -z "$LD" ]]; then
-  echo "unknown linker: pattern ld64-[0-9]* not found in 'ld -v' output";
-  test_result=${TEST_SKIP};
-elif [[ $LDVER -lt 224 ]]; then
-  echo "ld version is $LDVER, only 224 or above are supported";
+#!/usr/bin/env bash
+
+LDVER=$(ld -v 2>&1 | egrep -o "PROJECT:ld(64)?-[0-9]*" | sed -E "s/PROJECT:ld(64)?-//")
+
+# Extract the first 3 parts of an LD version number
+version () {
+    echo "$@" | awk -F. '{ printf("%d%03d%03d\n", $1,$2,$3); }'
+}
+
+if [ $(version "$LDVER") -lt $(version "224.0.0.0") ]; then
+  echo "ld version is $LDVER, only 224 or above are supported.";
   test_result=${TEST_SKIP};
 else
   test_result=${TEST_PASS};
