diff --git a/Changes b/Changes
index d5c9c4efea2c..0e76be3cf45b 100644
--- a/Changes
+++ b/Changes
@@ -22,6 +22,12 @@ Working version
 
 ### Code generation and optimizations:
 
+- #13262, #14074: fix performance issue on Apple Silicon macOS by emitting
+  `stlr` instead of `dmb ishld; str`.
+  (KC Sivaramakrishnan, report by François Pottier, analysis by Frédéric Bour,
+  Xavier Leroy, Miod Vallat, Gabriel Scherer and Stephen Dolan, review by Miod
+  Vallat, Vincent Laviron and Xavier Leroy)
+
 ### Standard library:
 
 - #13728: Add Sys.runtime_executable containing the full path (if available) to
